[10.x] Fix Postgres cache store failed to put exist cache in transaction

xdevor opened the pull request:
## Problem
When using Postgres as the Cache database store, 
an error occurs when performing put operation on an already existing cache key during a transaction.
The error is as follows: `SQLSTATE[25P02]: In failed sql transaction: 7 ERROR: current transaction is aborted, commands ignored until end of transaction block...`

## Steps To Reproduce
- Step 1: after setup postgres and .env file put the following code to laravel `routes/console.php`
```php
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\DB;

Artisan::command('test', function () {

    $store = Cache::store('database');

    DB::beginTransaction();

    $store->put('test-key', 'test-value');

    DB::commit();

});
```
- Step 2:  run `php artisan test`  **twice times** and the below error will show
<img width="1367" alt="Screen Shot 2023-11-10 at 10 39 39 PM" src="https://github.com/laravel/framework/assets/98752059/60db082c-1bf9-43b4-9490-9c2c58b9f197">

## Versions
- Postgres: v12 and v16
- PHP: 8.2
- Laravel: 10

xdevor commented on tests/Cache/CacheDatabaseStoreTest.php:
```diff
@@ -91,10 +91,24 @@ public function testValueIsUpdatedWhenInsertThrowsException()
+        $table->shouldReceive('upsert')->once()->with(['key' => 'prefixfoo', 'value' => base64_encode(serialize("\0")), 'expiration' => 61], 'key')->andReturnTrue();
+
+        $result = $store->put('foo', "\0", 60);
+        $this->assertTrue($result);
```
No problem, I've written an integration test to prove the bug exist and  also the integration test validity [here](https://github.com/xdevor/framework/pull/2). If the implementation of the test case is fine, I will push the commit here.

xdevor replied:
Okay, I have adjusted it to integration test

xdevor committed:
[10.x] Fix Postgres cache store failed to put exist cache in transaction

tpetry commented:
Wouldn't an upsert for all database drivers make the most sense? For existing keys they will have to execute two queries instead of just one with upserts.

xdevor commented:
@tpetry I 100% agree with your point, but the reason why I didn't do that in this bug fix PR is because I'm want to minimize side effect, also i'm not entirely sure if there has some reason that led to the previous implementation not using upsert.

tpetry commented:
@plehatron Was there a specific reason for not using upsert?

xdevor committed:
Test PostgresCacheStore integration test

xdevor commented:
If using the `upsert()` approach is confirmed, I would like to consider adjusting the operation of the `$store->add()` method in database cache store to also use `upsert()`, because the current implementation of `add()` has a problem that **it does not update expired cache**. However, **logically, an expired key should be treated as non-existent** cache and should be directly updated as a whole.

```php
    public function add($key, $value, $seconds)
    {
       ...
        try {
            return $this->table()->insert(compact('key', 'value', 'expiration'));
        } catch (QueryException) {
            return $this->table()
                ->where('key', $key)
                ->where('expiration', '<=', $this->getTime()) // <--- not update expired cache
                ->update([
                    'value' => $value,
                    'expiration' => $expiration,
                ]) >= 1;
        }
    }
```

xdevor committed:
[10.x] Improve put operation of database cache store

- Prevent non-rollback error query occur in transaction
- For existing keys they will have to execute two queries instead of just one with upserts

xdevor committed:
Add database cache store integration test

xdevor committed:
Chore: ad db cache store integration test to databases workflows

xdevor committed:
Fix SqlServerCacheStoreTest

xdevor committed:
Refactor db cache store integration test

xdevor committed:
Refactor db cache store integration test

xdevor committed:
Refactor style

xdevor commented:
It seems there's no issue with using `upsert` for databases supported by Laravel. I've adjusted it to use `upsert`.

crynobone committed:
wip

tpetry commented:
The `add` method has the same problem with PG when the `insert` fails because of a duplicate key. But honestly, I don't understand the code in the exception case. The logic seems to be different than described in the docblock.

plehatron commented:
> @plehatron Was there a specific reason for not using upsert?

Hey @tpetry, the original logic for `DatabaseStore::put` predates my work from https://github.com/laravel/framework/pull/26726 where I made sure the cache component is fully compliant with the PSR-16 SimpleCache interface. Maybe @taylorotwell has more insight into why upsert was not used then 🤷‍♂️.

xdevor commented:
> The `add` method has the same problem with PG when the `insert` fails because of a duplicate key. But honestly, I don't understand the code in the exception case. The logic seems to be different than described in the docblock.

@tpetry seems like the `add()` method is designed to ensure that the cache value is not overwritten when the cache exists. However, it is challenging to prevent race conditions without lock (and lock has risk), that's why the current implementation is as it is i think.

Perhaps the only solution for now is to delete expired caches either before each `add()` call or periodically delete expired cache.

tpetry commented:
> @tpetry seems like the `add()` method is designed to ensure that the cache value is not overwritten when the cache exists. However, it is challenging to prevent race conditions without lock (and lock has risk), that's why the current implementation like this.

But the logic in the catch-part is overwriting any value when the insert fails with an exception due to an existing value?

xdevor commented:
> But the logic in the catch-part is overwriting any value when the insert fails with an exception due to an existing value?

Because expired cache (`where('expiration', '<=', $this->getTime())` ) is same as not existing, and db will not remove the expired record automatically.

```php
        } catch (QueryException) {
            return $this->table()
                ->where('key', $key)
                ->where('expiration', '<=', $this->getTime()) // <--- only update expired cache, and expired cache is to be considered as not existing
                ->update([
                    'value' => $value,
                    'expiration' => $expiration,
                ]) >= 1;
        }
```

xdevor commented:
> Oh you're right. didn't look could enough. Still the `current transaction is aborted` would also happen if the `insert` fails, an `insertOrIgnore` needs to be used here.

@tpetry oh seems like possible ~ maybe implement like this to make sure it will not be silent while race condition occur, and call `get()` before insert  because `get()` will delete expired cache automatically
```php
        throw_if(
            is_null($this->table()->get($key))
            && $this->table()->insertOrIgnore(compact('key', 'value', 'expiration')) === 0,
            new \RuntimeException("Cache key '$key' has been set.")
        )
        return true;
```
